### ISSUE TITLE
LLVM backend producing js with Optimisation fails to run TypeError: r is not a function

### ISSUE THREAD

**Issue by jpharvey** [emscripten-core/emscripten#10675]
I have tested with latest version 1.39.10 and still see this problem.
It occurs during the initialization before we get into main

```
libthingview.js:29 Uncaught (in promise) TypeError: r is not a function
    at Lgj (libthingview.js:29)
    at Yyj (libthingview.js:101)
    at Module.___wasm_call_ctors (libthingview.js:115)
    at func (libthingview.js:115)
    at callRuntimeCallbacks (libthingview.js:115)
    at initRuntime (libthingview.js:115)
    at doRun (libthingview.js:115)
    at run (libthingview.js:115)
    at runCaller (libthingview.js:115)
    at removeRunDependency (libthingview.js:115)
```

Which comes from preamble.js :-

```
function initRuntime() {
#if STACK_OVERFLOW_CHECK
  checkStackCookie();
#endif
#if ASSERTIONS
  assert(!runtimeInitialized);
#endif
  runtimeInitialized = true;
  {{{ getQuoted('ATINITS') }}}
  callRuntimeCallbacks(__ATINIT__);
}
```
We are building with compiler flags 

`-O3 -s ASSERTIONS=2 -s DEMANGLE_SUPPORT=1 -DNDEBUG=1  --bind -s ALLOW_MEMORY_GROWTH=1 -s NO_EXIT_RUNTIME=1 -s FULL_ES2=1 --memory-init-file 0 -s ERROR_ON_UNDEFINED_SYMBOLS=1 -s DISABLE_EXCEPTION_CATCHING=0 -s EMBIND_STD_STRING_IS_UTF8=0 -s AGGRESSIVE_VARIABLE_ELIMINATION=1 -s DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0 --wasm=0
`

The same code built for webassembly (same flags except --wasm=0) works fine and non  of the sanitizers report any problems prior to getting into main.  I cant test with sanitizers and js that always fails to run, but i am not sure if it is expected to work.
SAFE_HEAP does not report any issues.
If i add --profiling the code runs without any problems.
Changing -O3 to -O1 or -O2 and the code runs fine.

With -Os i get a different manifestation of the problem 
```
Uncaught (in promise) Error: abort(stack overflow) at Error
    at jsStackTrace (libthingview.js:115)
    at stackTrace (libthingview.js:115)
    at abort (libthingview.js:115)
    at ___handle_stack_overflow (libthingview.js:115)
    at mSo (libthingview.js:17)
    at euq (libthingview.js:101)
    at Module.___wasm_call_ctors (libthingview.js:115)
    at func (libthingview.js:115)
    at callRuntimeCallbacks (libthingview.js:115)
    at initRuntime (libthingview.js:115)
    at abort (libthingview.js:115)
    at ___handle_stack_overflow (libthingview.js:115)
    at mSo (libthingview.js:17)
    at euq (libthingview.js:101)
    at Module.___wasm_call_ctors (libthingview.js:115)
    at func (libthingview.js:115)
    at callRuntimeCallbacks (libthingview.js:115)
    at initRuntime (libthingview.js:115)
    at doRun (libthingview.js:115)
    at run (libthingview.js:115)
abort @ libthingview.js:115
___handle_stack_overflow @ libthingview.js:115
mSo @ libthingview.js:17
euq @ libthingview.js:101
Module.___wasm_call_ctors @ libthingview.js:115
func @ libthingview.js:115
callRuntimeCallbacks @ libthingview.js:115
initRuntime @ libthingview.js:115
doRun @ libthingview.js:115
run @ libthingview.js:115
runCaller @ libthingview.js:115
removeRunDependency @ libthingview.js:115
receiveInstance @ libthingview.js:115
receiveInstantiatedSource @ libthingview.js:115
Promise.then (async)
instantiateArrayBuffer @ libthingview.js:115
instantiateAsync @ libthingview.js:115
createWasm @ libthingview.js:115
(anonymous) @ libthingview.js:115
```

If i just build our test code which uses smaller parts of the codebase they all run fine.
It is a large library that i cant provide as test code but i wondered whether providing the wasm file since presumably here the problem is likely to be in wasm2js would help?
Any further suggestions on how to help get this fixed?

**Comment by kripken** [comment]
This sounds a like an unknown wasm2js bug, ugh...

Too bad you can't share the source code, but if you can share the build, that would be enough for me to debug this I think. Specifically, please build with `-s WASM=2 --profiling`. That will emit both wasm2js output and normal wasm output. Please confirm that it works ok (and is easy for me to run, preferably in the shell) when run normally, and that it fails when running the wasm2js (to do that, just add `WebAssembly = undefined` at the top of the JS file - it will think the VM has no wasm support).

You can also try to add `EMCC_AUTODEBUG=1` to the env when compiling. If the problem still happens, providing that build as well (same as before) would be great (it adds tons of instrumentation that makes debugging easy, but it also sometimes changes optimization and bugs fail to reproduce).

**Comment by jpharvey** [comment]
I have shared the output of 3 different builds.
The generated js file in all cases has been modified to add WebAssembly=undefined.
These can be run with 
`node libthingview.js`
which will just load the library and run the initialization code which is what is failing at the moment.
Let me know if you need anything else or want me to try something.
Thanks

John

1)	thingview_O3.7z
https://drive.google.com/file/d/1rM2AualqVlrt5Ahf0fzahy_kssK9hFCG/view?usp=sharing
```
Compiler Flags:-
em++ --bind -s ALLOW_MEMORY_GROWTH=1 -s NO_EXIT_RUNTIME=1 -s FULL_ES2=1 --memory-init-file 0 -s ERROR_ON_UNDEFINED_SYMBOLS=1 -s DISABLE_EXCEPTION_CATCHING=0 -s EMBIND_STD_STRING_IS_UTF8=0 -s AGGRESSIVE_VARIABLE_ELIMINATION=1 -s DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0 -s ASM_JS=2 -s LEGACY_VM_SUPPORT=1 -s WASM=2 -s MIN_IE_VERSION=11 -O3 -s ASSERTIONS=2 -s DEMANGLE_SUPPORT=1 -DNDEBUG=1 -s DISABLE_EXCEPTION_CATCHING=0 --std=c++14  -o 
d:/scratch/pview_objects/vissrc_thingview1/js.o/js/ptc/thingview/libthingview.js @libthingview.tmp
```
webassembly runs, js fails with reported error
2)	thingview_O3_EMCC_AUTODEBUG.7z  https://drive.google.com/file/d/1QUPRz1srhpykdQ6geyxigEHtM6Kld71-/view?usp=sharing
Same compiler flags as _O3 but with EMCC_AUTODEBUG=1 and same results.
3)	Thingview_O3_profiling.7z 
https://drive.google.com/file/d/1nwsXfGM6j9sS1fTEE-giCX6G-z97VTKv/view?usp=sharing
Same compiler flags as _O3 but with –profiling, both wasm & js versions run ok.
Some wasm2js warnings:-
```
wasm2js: warning: global scope may be colliding with other scope: dg__DMat44__operator___28dg__DVec3_20const__29
wasm2js: warning: global scope may be colliding with other scope: bool_20std____2__operator___char_2c_20std____2__char_traits_char__20__28std____2__istreambuf_iterator_char_2c_20std____2__char_traits_char__20__20const__2c_20std____2__istreambuf_iterator_char_2c_20std____2__char_traits_char__20__20const__29
wasm2js: warning: global scope may be colliding with other scope: bool_20std____2__operator___wchar_t_2c_20std____2__char_traits_wchar_t__20__28std____2__istreambuf_iterator_wchar_t_2c_20std____2__char_traits_wchar_t__20__20const__2c_20std____2__istreambuf_iterator_wchar_t_2c_20std____2__char_traits_wchar_t__20__20const__29
wasm2js: warning: global scope may be colliding with other scope: $28anonymous_20namespace_29__itanium_demangle__OutputStream__operator___28_28anonymous_20namespace_29__itanium_demangle__StringView_29
wasm2js: warning: global scope may be colliding with other scope: dg__FMat33__operator___28dg__FMat33_20const__29
wasm2js: warning: global scope may be colliding with other scope: dg__DMat33__operator___28dg__DMat33_20const__29
wasm2js: warning: global scope may be colliding with other scope: cvu__operator__28cvu__Symbol_20const__2c_20cvu__Symbol_20const__29
```

**Comment by jpharvey** [comment]
@kripken it looks like your fix in #10682 fixes this issue (and makes the files smaller).  I have patched my copy of 1.39.10 with that fix. I will do some more testing tomorrow and close this issue assuming i dont find any problems. 
Thanks

John

**Comment by kripken** [comment]
Thanks @jpharvey ! Yeah, Ithink I understand the underlying issue now, so unless there is another bug you are hitting, that PR should fix things.

**Comment by jpharvey** [comment]
This was fixed by #10682 I have tested both -O3 & -Os both of which work now.

**Comment by kripken** [comment]
Great, thanks for confirming @jpharvey , and for the builds!
